pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -1026,6 +1044,166 @@ func registriesConfigIgnition(templateDir string, controllerConfig *mcfgv1.Contr
+		} else {
+			klog.Infof("default_runtime is already set in existing MachineConfigs for pool %s, skipping MachineConfig creation", pool.Name)
+			continue
+		}
+	}
+
+	// Create the config map for crio-default-container-runtime so we know that the crio-default-container-runtime MC has been created
+	if defaultContainerRuntimeCM == nil {
+		defaultContainerRuntimeCM = &v1.ConfigMap{}
+	}
+
+	defaultContainerRuntimeCM.Name = configMapName
+	defaultContainerRuntimeCM.Namespace = ctrl.namespace
+	if _, err := ctrl.kubeClient.CoreV1().ConfigMaps(ctrl.namespace).Create(context.TODO(), defaultContainerRuntimeCM, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("error creating %s config map: %w", configMapName, err)
+	}
+	return nil
+}
+
+func (ctrl *Controller) determineContainerRuntimeForFinalRuntimeConfig(pool *mcfgv1.MachineConfigPool) (mcfgv1.ContainerRuntimeDefaultRuntime, error) {
+	// Retrieve all ContainerRuntimeConfig objects for the pool
+	ctrcfgListAll, err := ctrl.client.MachineconfigurationV1().ContainerRuntimeConfigs().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error listing container runtime configs: %w", err)
+	}
+
+	var ctrcfgList []mcfgv1.ContainerRuntimeConfig
+	for _, ctrcfg := range ctrcfgListAll.Items {
+		selector, err := metav1.LabelSelectorAsSelector(ctrcfg.Spec.MachineConfigPoolSelector)
+		if err != nil {
+			return "", fmt.Errorf("invalid label selector: %w", err)
+		}
+		if selector.Empty() || !selector.Matches(labels.Set(pool.Labels)) {
+			continue
+		}
+		ctrcfgList = append(ctrcfgList, ctrcfg)
+	}
+	// Determine the final runtime configuration
+	var currentDefaultRuntime mcfgv1.ContainerRuntimeDefaultRuntime
+	var machineConfigs []*mcfgv1.MachineConfig
+
+	for _, ctrcfg := range ctrcfgList {
+		finalizers := ctrcfg.GetFinalizers()
+		for _, finalizer := range finalizers {
+			mcName := finalizer
+			mc, err := ctrl.client.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), mcName, metav1.GetOptions{})
+			if err != nil {
+				return "", fmt.Errorf("error getting MachineConfig %s: %w", mcName, err)
+			}
+			machineConfigs = append(machineConfigs, mc)
+		}
+	}
+
+	// Sort MachineConfigs in alphanumeric order
+	sort.Slice(machineConfigs, func(i, j int) bool {
+		return machineConfigs[i].Name < machineConfigs[j].Name
+	})
+
+	// Determine the final runtime configuration based on the last-applied MachineConfig
+	for i := len(machineConfigs) - 1; i >= 0; i-- {
+		mc := machineConfigs[i]
+		if mc.Spec.Config.Raw != nil {
+			ignConfig, err := ctrlcommon.ParseAndConvertConfig(mc.Spec.Config.Raw)
+			if err != nil {
+				return "", fmt.Errorf("error unmarshalling Ignition config: %w", err)
+			}
+			for _, file := range ignConfig.Storage.Files {
+				if file.Path == CRIODropInFilePathDefaultRuntime && file.Contents.Source != nil {
+					currentDefaultRuntime = mcfgv1.ContainerRuntimeDefaultRuntime(*file.Contents.Source)
```

Comment:
Maybe for safety reasons, we can add a check here to see if the contents fall under a set of expected values, in case we have some unexpected behaviour here

Reply:
Should I stick with runc/crun for now?

Reply:
I think that should be fine